File: src/Recursion/BackTracking/NKnights.java
package Recursion.BackTracking;

public class NKnights {
    public static void main(String[] args) {
        int n= 3;
        boolean[][] b = new boolean[n][n];
        NKnights(b,0,0,4);
    }
    static void NKnights(boolean[][] board, int row, int column, int target){
        if(target==0){
            display(board);
            System.out.println();
            return;
        }
        if(row==board.length-1 && column==board[0].length){
            return;
        }
        if(column==board[0].length){
            NKnights(board,row+1,0,target);
            return;
        }
        if(isSafe(board,row,column)){
            board[row][column] = true;
            NKnights(board,row,column+1,target-1);
            board[row][column] = false;
        }
        NKnights(board,row,column+1,target);
    }

    static boolean isSafe(boolean[][] board,int row, int column){
        if(isValid(board,row-2,column-1)){
            if(board[row-2][column-1]){
                return false;
            }
        }
        if(isValid(board,row-1,column-2)){
            if(board[row-1][column-2]){
                return false;
            }
        }
        if(isValid(board,row-1,column+2)){
            if(board[row-1][column+2]){
                return false;
            }
        }
        if(isValid(board,row-2,column+1)){
            if(board[row-2][column+1]){
                return false;
            }
        }
        return true;
    }
    static boolean isValid(boolean[][] board,int row,int column){
        if(row>=0 && row<board.length && column>=0 && column<board.length){
            return true;
        }
        return false;
    }
    static void display(boolean[][] board){
        for(boolean[] row:board){
            for(boolean elem:row){
                if(elem){
                    System.out.print("K ");
                }
                else{
                    System.out.print("X ");
                }
            }
            System.out.println();
        }
    }
}

File: pythonFiles/Recursion/BackTracking/NKnights.py
def NKnights(board,row,col,n):
    if n==0:
        print(board)
        return 1
    if row == len(board)-1 and col==len(board):
        return 0
    c = 0
    if col==len(board):
        c += NKnights(board,row+1,0,n)
    else:
        if isSafe(board,row,col):
            board[row][col] = "K"
            c += NKnights(board,row,col+1,n-1)
            board[row][col] = "."
        c += NKnights(board,row,col+1,n)
    return c

def isSafe(board,row,col):
    if(isValid(board,row-2,col+1) and board[row-2][col+1]=="K"):
        return False
    if(isValid(board,row-2,col-1) and board[row-2][col-1]=="K"):
        return False
    if(isValid(board,row-1,col-2) and board[row-1][col-2]=="K"):
        return False
    if(isValid(board,row-1,col+2) and board[row-1][col+2]=="K"):
        return False
    return True

def isValid(board,row,col):
    if row<len(board) and row>=0 and col<len(board) and col>=0:
        return True
    return False
n = 2
board = [['.' for i in range(n)] for i in range(n)]
print(NKnights(board,0,0,n))